### ISSUE TITLE
rpy2 error

### ISSUE THREAD

**Issue by Marius1311** [theislab/cellrank#92]
I finally got my rpy2 to run and I'm now really keen to get your gene_trends to run using R's `mgcv`. However, when I run 
```python
model = cr.ul.models.GamMGCVModel(adata, n_splines=7, sp=100)
cr.pl.gene_trends(adata, model=model, data_key='Ms',
                  genes=['Gng12', 'Pak3'],
                  time_key='latent_time',
                  weight_threshold=0.05, weight_scale=1, filter_data=True, dpi=25)
```
I get the error

```python
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-17-759d7e9f342a> in <module>
      8                   genes=['Gng12', 'Pak3'],
      9                   time_key='latent_time',
---> 10                   weight_threshold=0.05, weight_scale=1, filter_data=True, dpi=25)

~/Projects/cellrank/cellrank/plotting/_gene_trend.py in gene_trends(adata, model, genes, lineages, data_key, final, start_lineage, end_lineage, conf_int, same_plot, hide_cells, perc, lineage_cmap, abs_prob_cmap, cell_color, color, cell_alpha, lineage_alpha, size, lw, show_cbar, margins, sharey, figsize, dpi, ncols, n_jobs, backend, ext, suptitle, save, dirname, plot_kwargs, show_progres_bar, **kwargs)
    271         extractor=lambda modelss: {k: v for m in modelss for k, v in m.items()},
    272         show_progress_bar=show_progres_bar,
--> 273     )(lineages, start_lineage, end_lineage, **kwargs)
    274     logg.info("    Finish", time=start)
    275 

~/Projects/cellrank/cellrank/utils/_parallelize.py in wrapper(*args, **kwargs)
     98                 *((i, cs) if use_ixs else (cs,)), *args, **kwargs, queue=queue
     99             )
--> 100             for i, cs in enumerate(collections)
    101         )
    102 

~/miniconda3/envs/cellrank/lib/python3.6/site-packages/joblib/parallel.py in __call__(self, iterable)
   1002             # remaining jobs.
   1003             self._iterating = False
-> 1004             if self.dispatch_one_batch(iterator):
   1005                 self._iterating = self._original_iterator is not None
   1006 

~/miniconda3/envs/cellrank/lib/python3.6/site-packages/joblib/parallel.py in dispatch_one_batch(self, iterator)
    833                 return False
    834             else:
--> 835                 self._dispatch(tasks)
    836                 return True
    837 

~/miniconda3/envs/cellrank/lib/python3.6/site-packages/joblib/parallel.py in _dispatch(self, batch)
    752         with self._lock:
    753             job_idx = len(self._jobs)
--> 754             job = self._backend.apply_async(batch, callback=cb)
    755             # A job can complete so quickly than its callback is
    756             # called before we get here, causing self._jobs to

~/miniconda3/envs/cellrank/lib/python3.6/site-packages/joblib/_parallel_backends.py in apply_async(self, func, callback)
    207     def apply_async(self, func, callback=None):
    208         """Schedule a func to be run"""
--> 209         result = ImmediateResult(func)
    210         if callback:
    211             callback(result)

~/miniconda3/envs/cellrank/lib/python3.6/site-packages/joblib/_parallel_backends.py in __init__(self, batch)
    588         # Don't delay the application, to avoid keeping the input
    589         # arguments in memory
--> 590         self.results = batch()
    591 
    592     def get(self):

~/miniconda3/envs/cellrank/lib/python3.6/site-packages/joblib/parallel.py in __call__(self)
    254         with parallel_backend(self._backend, n_jobs=self._n_jobs):
    255             return [func(*args, **kwargs)
--> 256                     for func, args, kwargs in self.items]
    257 
    258     def __len__(self):

~/miniconda3/envs/cellrank/lib/python3.6/site-packages/joblib/parallel.py in <listcomp>(.0)
    254         with parallel_backend(self._backend, n_jobs=self._n_jobs):
    255             return [func(*args, **kwargs)
--> 256                     for func, args, kwargs in self.items]
    257 
    258     def __len__(self):

~/Projects/cellrank/cellrank/plotting/_utils.py in _fit(genes, lineage_names, start_lineages, end_lineages, queue, **kwargs)
    412             model = (
    413                 models[gene][ln]
--> 414                 .prepare(gene, ln, start_lineage=sc, end_lineage=ec, **kwargs)
    415                 .fit()
    416             )

~/Projects/cellrank/cellrank/utils/models/_models.py in fit(self, x, y, w, **kwargs)
    803         pandas2ri.activate()
    804 
--> 805         df = pandas2ri.py2rpy(
    806             pd.DataFrame(np.c_[self.x, self.y][use_ixs, :], columns=["x", "y"])
    807         )

AttributeError: module 'rpy2.robjects.pandas2ri' has no attribute 'py2rpy'
``` 

Could this be a  problem with my R or rpy2 version?

I'm using `r-base` version 3.5.1 and `rpy2` version 2.9.4.

**Comment by michalk8** [comment]
Mine rp2 version is `'3.2.6'`. I'd try upgrading it.

**Comment by Marius1311** [comment]
Okay, will do.

**Comment by Marius1311** [comment]
With conda, I can only update to `3.1.0`, see https://anaconda.org/search?q=rpy2

With that version, it still does not work. I will try uninstalling rpy2 and re-installing it using pip.

**Comment by Marius1311** [comment]
Okay, with pip, I get version `3.3.2`. This is something to note down - user installing with conda will always get an rpy2 version that raises the above error.

**Comment by Marius1311** [comment]
I get another error now though:

```python
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-19-759d7e9f342a> in <module>
      8                   genes=['Gng12', 'Pak3'],
      9                   time_key='latent_time',
---> 10                   weight_threshold=0.05, weight_scale=1, filter_data=True, dpi=25)

~/Projects/cellrank/cellrank/plotting/_gene_trend.py in gene_trends(adata, model, genes, lineages, data_key, final, start_lineage, end_lineage, conf_int, same_plot, hide_cells, perc, lineage_cmap, abs_prob_cmap, cell_color, color, cell_alpha, lineage_alpha, size, lw, show_cbar, margins, sharey, figsize, dpi, ncols, n_jobs, backend, ext, suptitle, save, dirname, plot_kwargs, show_progres_bar, **kwargs)
    271         extractor=lambda modelss: {k: v for m in modelss for k, v in m.items()},
    272         show_progress_bar=show_progres_bar,
--> 273     )(lineages, start_lineage, end_lineage, **kwargs)
    274     logg.info("    Finish", time=start)
    275 

~/Projects/cellrank/cellrank/utils/_parallelize.py in wrapper(*args, **kwargs)
     98                 *((i, cs) if use_ixs else (cs,)), *args, **kwargs, queue=queue
     99             )
--> 100             for i, cs in enumerate(collections)
    101         )
    102 

~/miniconda3/envs/cellrank/lib/python3.6/site-packages/joblib/parallel.py in __call__(self, iterable)
   1002             # remaining jobs.
   1003             self._iterating = False
-> 1004             if self.dispatch_one_batch(iterator):
   1005                 self._iterating = self._original_iterator is not None
   1006 

~/miniconda3/envs/cellrank/lib/python3.6/site-packages/joblib/parallel.py in dispatch_one_batch(self, iterator)
    833                 return False
    834             else:
--> 835                 self._dispatch(tasks)
    836                 return True
    837 

~/miniconda3/envs/cellrank/lib/python3.6/site-packages/joblib/parallel.py in _dispatch(self, batch)
    752         with self._lock:
    753             job_idx = len(self._jobs)
--> 754             job = self._backend.apply_async(batch, callback=cb)
    755             # A job can complete so quickly than its callback is
    756             # called before we get here, causing self._jobs to

~/miniconda3/envs/cellrank/lib/python3.6/site-packages/joblib/_parallel_backends.py in apply_async(self, func, callback)
    207     def apply_async(self, func, callback=None):
    208         """Schedule a func to be run"""
--> 209         result = ImmediateResult(func)
    210         if callback:
    211             callback(result)

~/miniconda3/envs/cellrank/lib/python3.6/site-packages/joblib/_parallel_backends.py in __init__(self, batch)
    588         # Don't delay the application, to avoid keeping the input
    589         # arguments in memory
--> 590         self.results = batch()
    591 
    592     def get(self):

~/miniconda3/envs/cellrank/lib/python3.6/site-packages/joblib/parallel.py in __call__(self)
    254         with parallel_backend(self._backend, n_jobs=self._n_jobs):
    255             return [func(*args, **kwargs)
--> 256                     for func, args, kwargs in self.items]
    257 
    258     def __len__(self):

~/miniconda3/envs/cellrank/lib/python3.6/site-packages/joblib/parallel.py in <listcomp>(.0)
    254         with parallel_backend(self._backend, n_jobs=self._n_jobs):
    255             return [func(*args, **kwargs)
--> 256                     for func, args, kwargs in self.items]
    257 
    258     def __len__(self):

~/Projects/cellrank/cellrank/plotting/_utils.py in _fit(genes, lineage_names, start_lineages, end_lineages, queue, **kwargs)
    412             model = (
    413                 models[gene][ln]
--> 414                 .prepare(gene, ln, start_lineage=sc, end_lineage=ec, **kwargs)
    415                 .fit()
    416             )

~/Projects/cellrank/cellrank/utils/models/_models.py in fit(self, x, y, w, **kwargs)
    787     ) -> "GamMGCVModel":
    788 
--> 789         from rpy2 import robjects
    790         from rpy2.robjects import pandas2ri, Formula
    791         from rpy2.robjects.packages import importr

~/miniconda3/envs/cellrank/lib/python3.6/site-packages/rpy2/robjects/__init__.py in <module>
     25 from . import conversion
     26 from . import vectors
---> 27 from . import language
     28 
     29 from rpy2.rinterface import (Sexp,

~/miniconda3/envs/cellrank/lib/python3.6/site-packages/rpy2/robjects/language.py in <module>
     14 _reval = ri.baseenv['eval']
     15 _parse = ri.parse
---> 16 _str2lang = ri.baseenv['str2lang']
     17 
     18 

~/miniconda3/envs/cellrank/lib/python3.6/site-packages/rpy2/rinterface_lib/conversion.py in _(*args, **kwargs)
     42 def _cdata_res_to_rinterface(function):
     43     def _(*args, **kwargs):
---> 44         cdata = function(*args, **kwargs)
     45         # TODO: test cdata is of the expected CType
     46         return _cdata_to_rinterface(cdata)

~/miniconda3/envs/cellrank/lib/python3.6/site-packages/rpy2/rinterface_lib/_rinterface_capi.py in _(*args, **kwargs)
    280 def _evaluated_promise(function):
    281     def _(*args, **kwargs):
--> 282         robj = function(*args, **kwargs)
    283         if _TYPEOF(robj) == openrlib.rlib.PROMSXP:
    284             robj = openrlib.rlib.Rf_eval(

~/miniconda3/envs/cellrank/lib/python3.6/site-packages/rpy2/rinterface_lib/sexp.py in __getitem__(self, key)
    353         # TODO: move check of R_UnboundValue to _rinterface
    354         if res == openrlib.rlib.R_UnboundValue:
--> 355             raise KeyError("'%s' not found" % key)
    356         return res
    357 

KeyError: "'str2lang' not found"
```

**Comment by Marius1311** [comment]
okay, but the above has nothing to do with your code - I get the same when I just run

```python
import rpy2
import rpy2.robjects as robjects
```

So this is a problem with my rpy2 setup, which does not seem to work again, having pip-installed. The above lines worked when I used the conda version. Maybe we can meet a bit earlier tomorrow, I share my screen and we finally try to sort this out?

**Comment by michalk8** [comment]
Sure, just tell me when.

**Comment by Marius1311** [comment]
Let's do 17:30

**Comment by michalk8** [comment]
Relevant: https://stackoverflow.com/questions/61491298/rpy2-latest-version-doesnt-work-on-ubuntu
What's your R version? I assume `< 3.6`, based on the comments from the above link.

**Comment by Marius1311** [comment]
That solved it! My R version was (and is) `3.5.1`. I downgraded `rpy2` to version `3.2.0`, as described in the link you posted, and that solved my issue! Awesome! I already managed to reproduce your `gene_trends` using `mgcv`!
